crates/wast/src/parser.rs

```diff
@@ -810,32 +901,137 @@ impl<'a> Cursor<'a> {
+    pub fn annotation(mut self) -> Option<(&'a str, Self)> {
+        let annot = loop {
+            match &self.parser.buf.tokens.get(self.cur)?.0 {
+                // assume tokens are the start of an annotation, and `None`
+                // will  get returned if we're wrong in `annotation_start`.
+                Source::Token(_) => break self.annotation_start()?,
+                // not skipping comments
+                Source::Comment(_) => return None,
+                // skipping whitespace ...
+                Source::Whitespace(_) => self.cur += 1,
+            }
+        };
+        self.cur += 2; // skip the `(` and the `@foo`
+        Some((annot, self))
+    }
+
+    /// After `annotation` is called, skips the rest of the annotation until the
+    /// end.
+    pub fn skip_annotation_internals(mut self) -> Option<Cursor<'a>> {
+        let mut depth = 1;
+        while depth > 0 {
+            match &self.parser.buf.tokens.get(self.cur)?.0 {
+                Source::Token(Token::LParen(_)) => depth += 1,
+                Source::Token(Token::RParen(_)) => depth -= 1,
+                _ => {}
+            }
+            self.cur += 1;
+        }
+        Some(self)
+    }
+
     fn advance_token(&mut self) -> Option<&'a Token<'a>> {
         let (token, next) = self.parser.buf.tokens.get(self.cur)?;
         match token {
-            // If the current token is a `Token`, then only advance ourselves
-            // one token further. That way if after this `Token` you want
-            // comments, we can find the comments!
-            Source::Token(t) => {
+            // If we're currently pointing at a token, and it's not the start
+            // of an annotation, then we return that token and advance
+            // ourselves one token...
+            Source::Token(t) if self.annotation_start().is_none() => {
                 self.cur += 1;
                 Some(t)
             }
 
-            // Otherwise this is a comment or whitespace token. In that case
-            // `next` points to the next `Token` token if `next` was listed.
+            // ... otherwise we need to skip the current token, and possibly
+            // more. Here we're skipping whitespace, comments, annotations, etc.
+            // Basically stuff that's intended to not be that relevant to the
+            // text format. This is a pretty common operation, though, and we
+            // may do it multiple times through peeks and such. As a result
+            // this is somewhat cached.
+            //
+            // Each token has a `Cell` pointing to the index of the next "real"
+            // token after it, or where to resume parsing. This is basically
+            // the return value (ish) of this function being cached. If the
+            // next pointer is 0 we haven't calculated it, otherwise it's
+            // either MAX (no next, we reached the end) or a number which
+            // points to the next token.
```

Comment:
I appreciate the comments here, but even better would be a newtype around `usize` (or even better `u32` or `u64`) that wrapped this up in an `enum`-ish + getters-that-return-`Option` API so that parsing code is just asking `n.is_end()` rather than open-coding `n == usize::MAX`.